core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/orm.go

```diff
@@ -34,16 +36,33 @@ func NewORM(chainID *big.Int, db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig) *
 	}
 }
 
-// InsertUpkeepState is idempotent and sets upkeep state values in db
-func (o *orm) InsertUpkeepState(state persistedStateRecord, qopts ...pg.QOpt) error {
+// BatchInsertUpkeepStates is idempotent and sets upkeep state values in db
+func (o *orm) BatchInsertUpkeepStates(state []persistedStateRecord, qopts ...pg.QOpt) error {
 	q := o.q.WithOpts(qopts...)
 
-	query := `INSERT INTO evm_upkeep_states (evm_chain_id, work_id, completion_state, block_number, inserted_at, upkeep_id, ineligibility_reason)
-	  VALUES ($1::NUMERIC, $2, $3, $4, $5, $6::NUMERIC, $7)
-	    ON CONFLICT (evm_chain_id, work_id)
-	    DO NOTHING`
+	fieldsPerRecord := 7
 
-	return q.ExecQ(query, o.chainID, state.WorkID, state.CompletionState, state.BlockNumber, state.InsertedAt, state.UpkeepID, state.IneligibilityReason)
+	valueStrings := make([]string, 0, len(state))
+	valueArgs := make([]interface{}, 0, len(state)*fieldsPerRecord)
+
+	for i, record := range state {
+		fieldOffset := i * fieldsPerRecord
+		valueStrings = append(valueStrings, fmt.Sprintf(
+			"($%d::NUMERIC, $%d, $%d, $%d, $%d, $%d::NUMERIC, $%d)",
+			fieldOffset+1, fieldOffset+2, fieldOffset+3, fieldOffset+4, fieldOffset+5, fieldOffset+6, fieldOffset+7,
+		))
+		valueArgs = append(valueArgs, o.chainID, record.WorkID, record.CompletionState, record.BlockNumber, record.InsertedAt, record.UpkeepID, record.IneligibilityReason)
+	}
+
+	query := fmt.Sprintf(`
+		INSERT INTO evm_upkeep_states (evm_chain_id, work_id, completion_state, block_number, inserted_at, upkeep_id, ineligibility_reason)
+		VALUES %s
+		ON CONFLICT (evm_chain_id, work_id)
+		DO NOTHING;
+	`, strings.Join(valueStrings, ", "))
```

Comment:
I think you might be able to do this more elegantly like https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/logpoller/orm.go#L187-L190